pkg/driver/kubernetes_driver.go

```diff
@@ -0,0 +1,366 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	k8sContainerName    = "invocation"
+	k8sFileSecretVolume = "files"
+)
+
+// KubernetesDriver is capable of running an invocation image in a Kubernetes cluster.
+type KubernetesDriver struct {
+	Namespace             string
+	ServiceAccountName    string
+	LimitCPU              resource.Quantity
+	LimitMemory           resource.Quantity
+	ActiveDeadlineSeconds int64
+	BackoffLimit          int32
+	SkipCleanup           bool
+	skipJobStatusCheck    bool
+	jobs                  batchclientv1.JobInterface
+	secrets               coreclientv1.SecretInterface
+	pods                  coreclientv1.PodInterface
+	deletionPolicy        metav1.DeletionPropagation
+	requiredCompletions   int32
+}
+
+// NewKubernetesDriver initializes a Kubernetes driver.
+func NewKubernetesDriver(namespace, serviceAccount string, conf *rest.Config) (*KubernetesDriver, error) {
+	driver := &KubernetesDriver{
+		Namespace:          namespace,
+		ServiceAccountName: serviceAccount,
+	}
+	driver.setDefaults()
+	err := driver.setClient(conf)
+	if err != nil {
+		return driver, err
+	}
+
+	return driver, nil
+}
+
+// Config returns the Kubernetes driver configuration options
+func (k *KubernetesDriver) Config() map[string]string {
+	return map[string]string{
+		"KUBE_NAMESPACE":  "Increase verbosity. true, false are supported values",
+		"SERVICE_ACCOUNT": "Always pull image, even if locally available (0|1)",
+		"KUBE_CONFIG":     "Absolute path to the kubeconfig file",
+		"MASTER_URL":      "Kubernetes master endpoint",
+	}
+}
+
+// SetConfig sets Kubernetes driver configuration
+func (k *KubernetesDriver) SetConfig(settings map[string]string) {
+	k.setDefaults()
+	k.Namespace = settings["KUBE_NAMESPACE"]
+	k.ServiceAccountName = settings["SERVICE_ACCOUNT"]
+
+	var kubeconfig string
+	if kpath := settings["KUBE_CONFIG"]; kpath != "" {
+		kubeconfig = kpath
+	} else if home := homeDir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	}
+
+	conf, err := clientcmd.BuildConfigFromFlags(settings["MASTER_URL"], kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	err = k.setClient(conf)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (k *KubernetesDriver) setDefaults() {
+	k.SkipCleanup = false
+	k.BackoffLimit = 0
+	k.ActiveDeadlineSeconds = 300
+	k.requiredCompletions = 1
+	k.deletionPolicy = metav1.DeletePropagationBackground
+}
+
+func (k *KubernetesDriver) setClient(conf *rest.Config) error {
+	coreClient, err := coreclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	batchClient, err := batchclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
+	}
+	k.jobs = batchClient.Jobs(k.Namespace)
+	k.secrets = coreClient.Secrets(k.Namespace)
+	k.pods = coreClient.Pods(k.Namespace)
+
+	return nil
+}
+
+// Run executes the operation inside of the invocation image
+func (k *KubernetesDriver) Run(op *driver.Operation) error {
```

Comment:
Personally, I think this method is too long and would benefit from factoring out helper methods/functions.

Reply:
Still long, but that shouldn't block this PR.